[WIP] Fix crash accessing godot physics space from AudioStreamPlayer3D whil…

lawnjelly opened the pull request:
…e paused

AudioStreamPlayer3D attempts to access the space_get_direct_state from the physics during update. This causes a crash with Godot physics when the scene tree is paused.

This fix does mean that areas will be ignored while paused (with Godot physics), although that is preferable to a crash.

It may be possible as a more in depth fix to allow access to areas, this would require a Godot physics contributor, and this PR will do as a stopgap solution.

Fixes #42108

Finally managed to compile on 4.x branch. But I can't currently test on 4.x as I'm getting a vulkan crash on startup (unrelated). I'll make the equivalent PR on 3.2 and leave someone else to test this I guess until I can get 4.x working.

<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

akien-mga commented on servers/physics/physics_server_sw.cpp:
```diff
@@ -192,11 +192,14 @@ real_t PhysicsServerSW::space_get_param(RID p_space, SpaceParameter p_param) con
-	ERR_FAIL_COND_V(!space, NULL);
-	ERR_FAIL_COND_V_MSG(!doing_sync || space->is_locked(), NULL, "Space state is inaccessible right now, wait for iteration or physics process notification.");
+
+	if (p_flag_error) {
```
I don't understand why this `p_flag_error` argument is added.

Not running `ERR_FAIL_COND_V(!space, NULL)` means crashing with a null pointer dereference when the condition is actually valid.

lawnjelly replied:
Ah yes you are right. I probably just need to mask out the doing_sync and is_locked.

Sorry I'm having problems with git at the moment, will update it as soon as I can work out how git works lol. :smile:

lawnjelly committed:
Fix crash accessing godot physics space from AudioStreamPlayer3D while paused

AudioStreamPlayer3D attempts to access the space_get_direct_state from the physics during update. This causes a crash with Godot physics when the scene tree is paused.

This fix does mean that areas will be ignored while paused (with Godot physics), although that is preferable to a crash.

It may be possible as a more in depth fix to allow access to areas, this would require a Godot physics contributor, and this PR will do as a stopgap solution.

Fixes #42108

madmiraal commented:
This doesn't actually fix #42108, it just hides the problem; so the "Fixes #42108" tag should be removed.

With regards the PR itself, there should be no need for a test to see whether the space state is available. The correct behaviour is the current behaviour i.e. print an error message if the space state is not available. If the space state isn't available the code shouldn't be even trying to access it, because it's not that it's temporarily unavailable, it's that it's always unavailable at this point.

The problem identified in #42132, is not that the space state may not be available, the problem is that it isn't available when the tree is paused. It should be, or at least it should be to objects that have the pause mode set to `PAUSE_MODE_PROCESS`.

So, I think the temporary "solution" should be limited to what was proposed in #42108: adding a `nullptr` test to the AudioStreamPlayer3D:
```
int areas = 0;
if (space_state) {
    areas = space_state->intersect_point(global_pos, sr, MAX_INTERSECT_AREAS, Set<RID>(), area_mask, false, true);
}
```
This will prevent the crash, but provide the required error message to aid with investigating and properly fixing the problem.

lawnjelly commented:
Ok, I'll change back to the original fix. :+1:

lawnjelly commented:
Leaving this PR for now temporarily till we have finalized #42137 .

In fact it may be better for someone else to cherry pick forward as I cannot run Godot 4.x currently.
